bundles/org.eclipse.packagedrone.utils.rpm/src/org/eclipse/packagedrone/utils/rpm/build/PayloadCoding.java

```diff
@@ -0,0 +1,195 @@
+/*******************************************************************************
+ * Copyright (c) 2016 IBH SYSTEMS GmbH and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     Red Hat Inc - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.packagedrone.utils.rpm.build;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.Optional;
+import java.util.zip.Deflater;
+
+import org.apache.commons.compress.compressors.bzip2.BZip2CompressorInputStream;
+import org.apache.commons.compress.compressors.bzip2.BZip2CompressorOutputStream;
+import org.apache.commons.compress.compressors.gzip.GzipCompressorInputStream;
+import org.apache.commons.compress.compressors.gzip.GzipCompressorOutputStream;
+import org.apache.commons.compress.compressors.gzip.GzipParameters;
+import org.apache.commons.compress.compressors.lzma.LZMACompressorInputStream;
+import org.apache.commons.compress.compressors.lzma.LZMACompressorOutputStream;
+import org.apache.commons.compress.compressors.xz.XZCompressorInputStream;
+import org.apache.commons.compress.compressors.xz.XZCompressorOutputStream;
+import org.apache.commons.compress.compressors.zstandard.ZstdCompressorInputStream;
+import org.apache.commons.compress.compressors.zstandard.ZstdCompressorOutputStream;
+import org.apache.commons.compress.compressors.zstandard.ZstdUtils;
+import org.eclipse.packagedrone.utils.rpm.deps.Dependency;
+import org.eclipse.packagedrone.utils.rpm.deps.RpmDependencyFlags;
+import org.tukaani.xz.LZMA2Options;
+
+public enum PayloadCoding
+{
+    NONE ( null ),
+    GZIP ( "gzip" ),
+    BZIP2 ( "bzip2" ),
+    LZMA ( "lzma" ),
+    XZ ( "xz" ),
+    ZSTD ( "zstd" );
+
+    private String coding;
+
+    private PayloadCoding ( final String coding )
+    {
+        this.coding = coding;
+    }
+
+    public String getCoding ()
+    {
+        return coding;
+    }
+
+    public static PayloadCoding fromCoding ( final String coding ) throws IOException
+    {
+        if ( coding == null )
+        {
+            return PayloadCoding.NONE;
+        }
+
+        for ( PayloadCoding payloadCoding : PayloadCoding.values () )
+        {
+            if ( coding.equals ( payloadCoding.getCoding () ) )
+            {
+                return payloadCoding;
+            }
+        }
+
+        throw new IOException ( String.format ( "Unknown coding: %s", coding ) );
+    }
+
+    public static Optional<Dependency> getDependency ( final PayloadCoding payloadCoding ) throws IOException
```

Comment:
Consider whether we should move the following methods into an interface and have different implementations for each `PayloadCoding` value. At least we should determine if these are the right methods.

Reply:
I think you could add them to the enum itself.

~~~java
enum PayloadCoding {
  private final Optional<Dependency> dependency;

  private PayloadCoding(final Dependency dependency) {
    this.dependency = Optional.ofNullable(dependency);
  }

  public Optional<Dependency> getDependency () {
    return this.dependency;
  }

  NONE(null),
  …
  XZ(new Dependency ( "PayloadIsXz", "5.2-1", RpmDependencyFlags.LESS, RpmDependencyFlags.EQUAL, RpmDependencyFlags.RPMLIB )),
  …
  ;
}
~~~

Reply:
Please forget my previous comment. You already worked towards that idea.

> Consider whether we should move the following methods into an interface and have different implementations for each PayloadCoding value. At least we should determine if these are the right methods.

Yes, those are method which should go into an interface, and then their implementations.

I think that once this interface and the implementations are extracted, the actual `enum` may be irrelevant. As all the necessary information is inside the implementation of each coding.